README.md
# Admin Commenters Comments Count

A plugin for WordPress that displays a count of each commenter's total number of comments (linked to those comments) next to their name on any admin page.

This plugin can be found in the WordPress Plugin Directory: https://wordpress.org/plugins/admin-commenters-comments-count/

## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/admin-commenters-comments-count/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/admin-commenters-comments-count/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/admin-commenters-comments-count/). If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/admin-commenters-comments-count/reviews/) for it in the WordPress Plugin Directory or considerating a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## _(in-progress)_
* Fix: Change `__wakeup()` method visibility from `private` to `public` to avoid warnings under PHP8
* Fix: Throw an error when attempting to unserialize an instance of the class to actually prevent it from happening

## 1.9.5 _(2021-04-22)_
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)
* New: Add a few more possible TODO items

## 1.9.4 _(2020-09-13)_
* Change: Convert to use of strict equality checks
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Note compatibility through WP 5.5+
* Change: Tweak inline function documentation
* Change: Update list of TODO items to add some considerations to an existing item, fix a type, change sublist syntax
* Unit tests:
    * New: Add tests for `add_user_column()`, `enqueue_admin_css()`, `handle_column_data()`

## 1.9.3 _(2020-06-03)_
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add to it)
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and remove commented-out code)
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* New: Unit tests: Add test and data provider for hooking actions and filters

## 1.9.2 _(2019-12-07)_
* Fix: Correct typo in GitHub URL
* Unit test:
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * Change: Update expected frontend output to include additional "ugc" (user generated content) value now included for `rel="nofollow"` attribute of commenter links
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)

## 1.9.1 _(2019-04-05)_
* Change: Initialize plugin on `plugins_loaded` action instead of on load
* Change: Merge `do_init()` into `init()`
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* Change: Unit tests: Specify hook priority when testing via `has_filter()`
* Change: Note compatibility through WP 5.1+
* Change: Update copyright date (2019)
* Change: Update License URI to be HTTPS
* Change: Split paragraph in README.md's "Support" section into two

## 1.9 _(2017-11-06)_
* Bugfix: Explicitly set comment count font color to avoid style conflict with Akismet that resulted in gray text on dark gray background
* Bugfix: Disable Akismet's version of the functionality since it is duplicative and interferes with author section layout
* Change: Omit unnecessary `wp_register_style()` and instead provide all arguments to `wp_enqueue_style()`
* Harden: Use 'esc_like()` on the pingback/trackback URL prior to use in queries
* Bugfix: Use proper existing variable when searching for other pingbacks/trackbacks
* Change: Omit unnecessary appending of '%' to author_url value in call to `get_comments_count()`
* New: Add README.md
* Change: Add GitHub link to readme
* Change: Note compatibility through WP 4.9+
* Change: Update copyright date (2018)
* Change: Minor whitespace tweaks in unit test bootstrap

## 1.8 _(2017-03-04)_
* Change: Adopt WP core style of showing pending comments in a red circle superscript to comments bubble icon
* Change: Show comments bubble in "Comments" column of user listings instead of plain integer
* Change: Don't link comments bubble when there are zero comments since approved and pending comments are available as separate links
* Change: Extract comments bubble markup generation from `comment_author()` into new `get_comments_bubble()`
* Change: Remove support for pre-WP 4.3 markup
* Change: Prevent object instantiation of the class
* Change: Use `sprintf()` to produce markup rather than concatenating various strings, function calls, and variables
* Change: Update unit test bootstrap
    * Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable
    * Enable more error output for unit tests
* Change: Note compatibility through WP 4.7+
* Change: Remove support for WordPress older than 4.6 (should still work for earlier versions back to WP 4.3)
* Change: Update readme.txt content and formatting
* Change: Update copyright date (2017)
* Change: Update screenshots
* New: Add LICENSE file

## 1.7 _(2016-01-11)_
* Bugfix: Fix bug preventing display of values for other custom columns in users table
* Add: Memoize commenter counts so they aren't re-queried more than once per page load
* Add: Add support for language packs by loading textdomain and explicit using it
* Add: Add inline docs for class variables
* Add: Add public method `reset_cache()`
* Add: Add private variable `$memoized`
* Change: Note compatibility through WP 4.4+
* Change: Explicitly declare methods in unit tests as public
* Change: Update copyright date (2016)
* Add: Create empty index.php to prevent files from being listed if web server has enabled directory listings

## 1.6 _(2015-09-19)_
* Bugfix: Add support for changes in WP 4.3 (to fix display of comment bubble background).
* Change: Minor inline documentation spacing tweaks.
* Add: Add 'Text Domain' field to plugin header.
* Change: Note compatibility through WP 4.3+.

## 1.5 _(2015-02-05)_
* Add `is_admin()` check to `comment_author()`
* Always load class rather than just in the admin
* Add to, and improve, unit tests
* Add screenshot showing 'Comments' column in user listing
* Curly-braced variables used in strings
* Note compatibility through WP 4.1+
* Update copyright date (2015)

## 1.4 _(2014-08-30)_
* Add 'Comments' column to admin user listing with linked count of that user's comments
* Modify markup output to accommodate changes made in WP 3.9

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Admin_Commenters_Comments_Count
 */


$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
        $_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
        require dirname( dirname( __FILE__ ) ) . '/admin-commenters-comments-count.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


admin-commenters-comments-count.php
<?php
/**
 * Plugin Name: Admin Commenters Comments Count
 * Version:     1.9.5
 * Plugin URI:  https://coffee2code.com/wp-plugins/admin-commenters-comments-count/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: admin-commenters-comments-count
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Displays a count of each commenter's total number of comments (linked to those comments) next to their name on any admin page.
 *
 * Compatible with WordPress 4.6 through 5.7+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/admin-commenters-comments-count/
 *
 * @package Admin_Commenters_Comments_Count
 * @author  Scott Reilly
 * @version 1.9.5
 */

/*
	Copyright (c) 2009-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_AdminCommentersCommentsCount' ) ) :

class c2c_AdminCommentersCommentsCount {

	/**
	 * The field name.
	 *
	 * Changing this requires changing .css and .js files.
	 *
	 * @var string
	 * @access private
	 */
	private static $field       = 'commenters_count';

	/**
	 * The field title.
	 *
	 * @var string
	 * @access private
	 */
	private static $field_title = '';

	/**
	 * Memoized array of commenter counts.
	 *
	 * @var string
	 * @access private
	 *
	 * @since 1.7
	 */
	private static $memoized    = array();

	/**
	 * Prevent instantiation.
	 *
	 * @since 1.8
	 */
	private function __construct() {}

	/**
	 * Prevent unserializing an instance.
	 *
	 * @since 1.8
	 * @since 1.8.3 Changed method visibility from private to public and throw exception if invoked.
	 */
	public function __wakeup() {
		/* translators: %s: Name of plugin class. */
		throw new Error( sprintf( __( '%s cannot be unserialized.', 'admin-commenters-comments-count' ), __CLASS__ ) );
	}

	/**
	 * Returns version of the plugin.
	 *
	 * @since 1.1.4
	 */
	public static function version() {
		return '1.9.5';
	}

	/**
	 * Initializer
	 */
	public static function init() {
		// Set translatable and filterable strings
		self::$field_title = __( 'Comments', 'admin-commenters-comments-count' );

		// Load textdomain.
		load_plugin_textdomain( 'admin-commenters-comments-count' );

		// Register hooks
		add_filter( 'comment_author',             array( __CLASS__, 'comment_author'          )        );
		add_filter( 'get_comment_author_link',    array( __CLASS__, 'get_comment_author_link' )        );
		add_action( 'admin_enqueue_scripts',      array( __CLASS__, 'enqueue_admin_css'       )        );
		add_filter( 'manage_users_columns',       array( __CLASS__, 'add_user_column'         )        );
		add_filter( 'manage_users_custom_column', array( __CLASS__, 'handle_column_data'      ), 10, 3 );

		// Disable Akismet's version of this functionality.
		add_filter( 'akismet_show_user_comments_approved', '__return_false' );
	}

	/**
	 * Resets cached data.
	 *
	 * @since 1.7
	 */
	public static function reset_cache() {
		self::$memoized = array();
	}

	/**
	 * Enqueues stylesheets if the user has admin expert mode activated.
	 *
	 * @since 1.3
	 */
	public static function enqueue_admin_css() {
		wp_enqueue_style( __CLASS__ . '_admin', plugins_url( 'assets/admin.css', __FILE__ ), array(), self::version() );
	}

	/**
	 * Adds a column for the count of user comments.
	 *
	 * @since 1.4
	 *
	 * @param  array $users_columns Array of user column titles.
	 *
	 * @return array The $posts_columns array with the user comments count column's title added.
	 */
	public static function add_user_column( $users_columns ) {
		$users_columns[ self::$field ] = self::$field_title;

		return $users_columns;
	}

	/**
	 * Outputs the markup for a linked count of the user's comments.
	 *
	 * @since 1.4
	 *
	 * @param string $output      Custom column output. Default empty.
	 * @param string $column_name Column name.
	 * @param int    $user_id     ID of the currently-listed user.
	 * @return string
	 */
	public static function handle_column_data( $output, $column_name, $user_id ) {
		if ( self::$field !== $column_name ) {
			return $output;
		}

		$user = get_user_by( 'id', $user_id );

		list( $comment_count, $pending_count ) = self::get_comments_count( 'comment_author_email', $user->user_email, 'comment', $user->ID );

		$msg = sprintf( _n( '%d comment', '%d comments', $comment_count, 'admin-commenters-comments-count' ), $comment_count );

		return self::get_comments_bubble( $user->user_email, $comment_count, $pending_count, $msg, false );
	}

	/**
	 * Gets the count of comments and pending comments for the given user.
	 *
	 * @since 1.4
	 *
	 * @param  string $field   The comment field value to search. One of: comment_author, comment_author_email, comment_author_IP, comment_author_url, user_id
	 * @param  string $value   The value of the field to check for.
	 * @param  string $type    Optional. comment_type value. Default 'comment'.
	 * @param  int    $user_id Optional. The user ID. This is searched for as an OR to the $field search. If set, forces $type to be 'comment'. Default 0.
	 *
	 * @return array  Array of comment count and pending count.
	*/
	public static function get_comments_count( $field, $value, $type = 'comment', $user_id = 0 ) {
		global $wpdb;

		if ( ! in_array( $field, array( 'comment_author', 'comment_author_email', 'comment_author_IP', 'comment_author_url', 'user_id' ) ) ) {
			$field = 'comment_author_email';
		}

		// Check if count has already been memoized.
		$memoize_key = "{$field}_{$value}_{$type}_" . (int) $user_id;
		if ( isset( self::$memoized[ $memoize_key ] ) && self::$memoized[ $memoize_key ] ) {
			return self::$memoized[ $memoize_key ];
		}

		// Query for counts.
		if ( $user_id && 'user_id' != $field ) {
			$query = "SELECT COUNT(*) FROM {$wpdb->comments} WHERE ( {$field} = %s OR user_id = %d ) AND comment_approved = %d";
			$comment_count = $wpdb->get_var( $wpdb->prepare( $query, $value, $user_id, 1 ) );
			$pending_count = $wpdb->get_var( $wpdb->prepare( $query, $value, $user_id, 0 ) );
		} elseif ( 'comment' == $type ) {
			$query = "SELECT COUNT(*) FROM {$wpdb->comments} WHERE {$field} = %s AND comment_approved = %d";
			$comment_count = $wpdb->get_var( $wpdb->prepare( $query, $value, 1 ) );
			$pending_count = $wpdb->get_var( $wpdb->prepare( $query, $value, 0 ) );
		} elseif ( 'pingback' == $type || 'trackback' == $type ) {
			$query = "SELECT COUNT(*) FROM {$wpdb->comments} WHERE comment_author_url LIKE %s AND comment_type = %s AND comment_approved = %d";
			$comment_count = $wpdb->get_var( $wpdb->prepare( $query, $wpdb->esc_like( $value ) . '%', $type, 1 ) );
			$pending_count = $wpdb->get_var( $wpdb->prepare( $query, $wpdb->esc_like( $value ) . '%', $type, 0 ) );
		} else {
			$comment_count = $pending_count = 0;
		}

		// Return value is array of comment count and pending count.
		$return = array( (int) $comment_count, (int) $pending_count );

		// Memoize the counts being returned.
		self::$memoized[ $memoize_key ] = $return;

		return $return;
	}

	/**
	 * Returns the URL for the admin listing of comments for the given email
	 * address.
	 *
	 * @since 1.4
	 *
	 * @param  string $email Email address.
	 *
	 * @return string
	 */
	public static function get_comments_url( $email ) {
		return add_query_arg( 's', urlencode( $email ), admin_url( 'edit-comments.php' ) );
	}

	/**
	 * Returns the comment author link for the specified author along with the
	 * markup indicating the number of times the comment author has commented
	 * on the site.  The comment count links to a listing of all of that
	 * person's comments.
	 *
	 * Commenters are identified by the email address they provided when
	 * commenting.
	 *
	 * @param string $author_name Name of the comment author.
	 *
	 * @return string Comment author link plus linked comment count markup.
	 */
	public static function comment_author( $author_name ) {
		if ( ! is_admin() ) {
			return $author_name;
		}

		global $comment;
		$type = get_comment_type();

		if ( 'comment' === $type ) {
			$author_email = $comment->comment_author_email;
			$author_name  = $comment->comment_author;
			if ( ! $author_email ) {
				$field = 'comment_author';
				$value = $author_name;
			} else {
				$field = 'comment_author_email';
				$value = $author_email;
			}
			list( $comment_count, $pending_count ) = self::get_comments_count( $field, $value, $type );
			$msg = sprintf( _n( '%d comment', '%d comments', $comment_count, 'admin-commenters-comments-count' ), $comment_count );
		} elseif ( 'pingback' === $type || 'trackback' === $type ) {
			$author_url = $comment->comment_author_url;
			// Want to get the root domain and not use the exact pingback/trackback source link
			$parsed_url = parse_url( $author_url );
			$author_url = $parsed_url['scheme'] . '://' . $parsed_url['host'];
			list( $comment_count, $pending_count ) = self::get_comments_count( 'comment_author_url', $author_url, $type );
			$author_email = $author_url;
			/* Translators: sorry, but I'm not supplying explicit translation strings for all possible other comment types.
			   You can at least expect '%d trackback', '%d trackbacks', '%d pingback' and '%d pingbacks' */
			$msg = sprintf( _n( '%d %s', '%d %ss', $comment_count, 'admin-commenters-comments-count' ), $comment_count, $type );
		} else {
			return $author_name;
		}

		// If appearing on the dashboard, then don't need to break out of
		// pre-existing <strong> tags.
		$screen = get_current_screen();
		$is_dashboard = $screen && 'dashboard' === $screen->id;

		$html = $is_dashboard ? '' : '</strong>';

		$html .= self::get_comments_bubble( $author_email, $comment_count, $pending_count, $msg );

		$html .= $is_dashboard ? '' : '<strong>';
		$html .= $author_name;

		return $html;
	}

	/**
	 * Returns the markup for the comments bubble for the given email address with
	 * with the provided counts.
	 *
	 * @since 1.8
	 *
	 * @param string $author_email       Comment author email address.
	 * @param int    $comment_count      The number of comments for the email
	 *                                   address.
	 * @param int    $pending_count      The number of pending comments for the
	 *                                   email address.
	 * @param string $msg                String to use as title attribute for
	 *                                   comment bubble.
	 * @param bool   $no_comments_bubble Should the comment bubble be shown if the
	 *                                   email address has no comments?
	 *
	 * @return string
	 */
	public static function get_comments_bubble( $author_email, $comment_count, $pending_count, $msg = '', $no_comments_bubble = true ) {
		$html = '';

		$url = ( $comment_count + $pending_count ) > 0 ? self::get_comments_url( $author_email ) : '#';

		$comment_str = sprintf(
			_n( '%s comment', '%s comments', $comment_count, 'admin-commenters-comments-count' ),
			number_format_i18n( $comment_count )
		);

		$pending_class = $pending_count ? '' : ' author-com-count-no-pending';

		if ( ! $no_comments_bubble && ! $comment_count && ! $pending_count ) {
			return sprintf( '<span aria-hidden="true">—</span><span class="screen-reader-text">%s</span>',
				__( 'No comments', 'admin-commenters-comments-count' )
			);
		}

		$html .= '<span class="column-response"><span class="post-com-count-wrapper post-and-author-com-count-wrapper author-com-count' . $pending_class . "\">\n";

		$comments_number = number_format_i18n( $comment_count );

		if ( $comment_count ) {
			$html .= sprintf(
				'<a href="%s" title="%s" class="post-com-count post-com-count-approved">
					<span class="comment-count-approved" aria-hidden="true">%s</span>
					<span class="screen-reader-text">%s</span>
				</a>',
				esc_url( add_query_arg( 'comment_status', 'approved', $url ) ),
				esc_attr( $msg ),
				$comments_number,
				$comment_str
			);
		} else {
			$html .= sprintf(
				'<span class="post-com-count post-com-count-no-comments" title="%s"><span class="comment-count comment-count-no-comments" aria-hidden="true">%s</span><span class="screen-reader-text">%s</span></span>',
				esc_attr( $msg ),
				$comments_number,
				$pending_count ? __( 'No approved comments', 'admin-commenters-comments-count' ) : __( 'No comments', 'admin-commenters-comments-count' )
			);
		}

		$pending_comments_number = number_format_i18n( $pending_count );
		$pending_phrase = sprintf( _n( '%s pending comment', '%s pending comments', $pending_count, 'admin-commenters-comments-count' ), $pending_comments_number );
		if ( $pending_count ) {
			$html .= sprintf(
				'<a href="%s" class="post-com-count post-com-count-pending"><span class="comment-count-pending" aria-hidden="true">%s</span><span class="screen-reader-text">%s</span></a>',
				esc_url( add_query_arg( 'comment_status', 'moderated', $url ) ),
				$pending_comments_number,
				$pending_phrase
			);
		} else {
			$html .= sprintf(
				'<span class="post-com-count post-com-count-pending post-com-count-no-pending"><span class="comment-count comment-count-no-pending" aria-hidden="true">%s</span><span class="screen-reader-text">%s</span></span>',
				$pending_comments_number,
				$comment_count ? __( 'No pending comments', 'admin-commenters-comments-count' ) : __( 'No comments', 'admin-commenters-comments-count' )
			);
		}

		$html .= "</span></span>";

		return $html;
	}

	/**
	 * Filter for WP's get_comment_author_link() that returns the value of
	 * comment_author() when in the admin.
	 *
	 * @param string $author_link Author link
	 *
	 * @return string Modified author link
	 */
	public static function get_comment_author_link( $author_link ) {
		if ( ! is_admin() ) {
			return $author_link;
		}

		return self::comment_author( get_comment_author() );
	}
} // end c2c_AdminCommentersCommentsCount

add_action( 'plugins_loaded', array( 'c2c_AdminCommentersCommentsCount', 'init' ) );

endif; // end if !class_exists()
